File: solutions/ocaml/perfect-numbers/1/perfect_numbers.ml
open Base

(* ---- Helpers ---- *)

let integer_square_root number =
    number
    |> Int.to_float
    |> Float.sqrt
    |> Float.to_int

let is_divisor number candidate =
    Int.rem number candidate = 0

let get_divisors_up_to_limit number limit =
    let upper_bound = limit + 1 in
    upper_bound
    |> List.range 1
    |> List.filter ~f:(is_divisor number)

let expand_pair_for_divisor number divisor =
    let paired_divisor = number / divisor in
    if paired_divisor = divisor
        then [divisor]  (* perfect square *)
        else [divisor; paired_divisor]

let expand_divisor_pair number divisor =
    if divisor = 1
        then [1]
        else expand_pair_for_divisor number divisor

let is_not_self_divisor number divisor =
    divisor <> number

let get_proper_divisors number =
    number
    |> integer_square_root
    |> get_divisors_up_to_limit number
    |> List.concat_map ~f:(expand_divisor_pair number)
    |> List.filter ~f:(is_not_self_divisor number)

let proper_divisors number =
    if number <= 1
        then []
        else get_proper_divisors number

let sum_of_numbers numbers =
    List.fold numbers ~init:0 ~f:(+)

(* ---- Classification ---- *)

let classify_by_divisor_sum number =
    let divisors = proper_divisors number in
    let sum = sum_of_numbers divisors in
    if sum = number
        then Ok "perfect"
        else if sum > number
            then Ok "abundant"
            else Ok "deficient"

let classify number =
    if number <= 0
        then Error "Classification is only possible for positive integers."
        else classify_by_divisor_sum number

File: solutions/python/perfect-numbers/1/perfect_numbers.py
CLASSIFICATION_ERROR_MESSAGE = (
    "Classification is only possible for positive integers."
)

PERFECT = "perfect"
ABUNDANT = "abundant"
DEFICIENT = "deficient"

FIRST_PROPER_DIVISOR = 1
FIRST_DIVISOR_TO_TEST = 2

def validate_number(number):
    if number <= 0:
        raise ValueError(CLASSIFICATION_ERROR_MESSAGE)

def get_aliquot_sum(number):
    return sum(_get_proper_divisors(number))

def _get_proper_divisors(number):
    if number == FIRST_PROPER_DIVISOR:
        return []

    divisors = [FIRST_PROPER_DIVISOR]
    divisor = FIRST_DIVISOR_TO_TEST
    while divisor ** 2 <= number:
        if _is_divisor(number, divisor):
            divisors.extend(_get_divisor_pair(number, divisor))

        divisor += 1

    return divisors

def _is_divisor(number, candidate):
    return number % candidate == 0

def _get_divisor_pair(number, divisor):
    paired_divisor = number // divisor
    if paired_divisor == divisor:
        return [divisor]

    return [divisor, paired_divisor]

def classify(number):
    validate_number(number)
    aliquot_sum = get_aliquot_sum(number)

    if aliquot_sum == number:
        return PERFECT

    if aliquot_sum > number:
        return ABUNDANT

    return DEFICIENT